components/tikv_util/src/quota_limiter.rs

```diff
@@ -0,0 +1,200 @@
+
+// The cpu time is not a real statistics, only part of the processing logic is
+// taken into account, so it needs to be multiplied by a factor.
+// transfer milli cpu to micro cpu
+//
+// TODO: Don't adjusted based on experience.
+const CPU_TIME_FACTOR: f64 = 0.8;
+
+// To avoid long tail latency.
+const MAX_QUOTA_DELAY: Duration = Duration::from_secs(1);
+
+// Quota limiter allows users to obtain stable performance by increasing the
+// completion time of tasks through restrictions of different metrics.
+#[derive(Debug)]
+pub struct QuotaLimiter {
+    cputime_limiter: Limiter,
+    write_bandwidth_limiter: Limiter,
+    read_bandwidth_limiter: Limiter,
+}
+
+// Throttle should be used by method `delay()` after `QuotaLimiter` consume.
+pub struct Throttle(Duration);
+
+impl Throttle {
+    pub fn is_throttled(&self) -> bool {
+        !self.0.is_zero()
+    }
+
+    pub fn ref_delay(&self) -> &Duration {
+        &self.0
+    }
+
+    pub async fn delay(&self) {
+        GLOBAL_TIMER_HANDLE
+            .delay(std::time::Instant::now() + self.0)
+            .compat()
+            .await
+            .unwrap();
+    }
+}
+
+impl Default for QuotaLimiter {
+    fn default() -> Self {
+        Self {
+            cputime_limiter: Limiter::new(f64::INFINITY),
+            write_bandwidth_limiter: Limiter::new(f64::INFINITY),
+            read_bandwidth_limiter: Limiter::new(f64::INFINITY),
+        }
+    }
+}
+
+impl QuotaLimiter {
+    // 1000 millicpu equals to 1vCPU, 0 means unlimited
+    pub fn new(
+        cpu_quota: usize,
+        write_bandwidth: ReadableSize,
+        read_bandwidth: ReadableSize,
+    ) -> Self {
+        let cputime_limiter = if cpu_quota == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(cpu_quota as f64 * CPU_TIME_FACTOR * 1000_f64)
+        };
+
+        let write_bandwidth_limiter = if write_bandwidth.0 == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(write_bandwidth.0 as f64)
+        };
+
+        let read_bandwidth_limiter = if read_bandwidth.0 == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(read_bandwidth.0 as f64)
+        };
+
+        Self {
+            cputime_limiter,
+            write_bandwidth_limiter,
+            read_bandwidth_limiter,
+        }
+    }
+
+    // record info after write requests finished and return the suggested delay duration.
+    pub fn consume_write(&self, bytes: usize, cpu_time: Duration) -> Throttle {
+        let cpu_dur = self
+            .cputime_limiter
+            .consume_duration(cpu_time.as_micros() as usize);
+
+        let bw_dur = if bytes > 0 {
+            self.write_bandwidth_limiter.consume_duration(bytes)
+        } else {
+            Duration::ZERO
+        };
+
+        let max_dur = std::cmp::max(cpu_dur, bw_dur);
+        let should_delay = if max_dur > cpu_time {
+            max_dur - cpu_time
```

Comment:
why

Reply:
this is a rate limiter to control bandwidth, but for cpu, we need small adjustment. If cpu used 500ms but quota to 1000ms, we don't need limit because cpu already waste real 500ms as its cost. So it only need to compensate for the insufficient buckets.

Reply:
this is the exact argument point between xuqi and me, for now only sync sleep actually waste real time and for async mode, I prefer call the `forefront-cpu-time` a indicator, it's not represent real cpu time. You can refer 6th pic in the description

Reply:
anyway in async mode both calculation methods are not fully correct. But current method is more in line with the description of the token bucket algorithm.

Reply:
btw, I would like to express that even use the old method, it is still impossible to compensate correctly in asynchronous mode, and now we are using a fixed number to multiply the config parameter, which is completely from our experience.

Reply:
understand